### PR TITLE
Stop copying LogicalPlan and Exprs in `PushDownFilter` (4%-6% faster planning)

### PR DESCRIPTION
**Pull request by alamb**
~Draft as it depends on~
*  builds on https://github.com/apache/datafusion/pull/10437 
* Builds on https://github.com/apache/datafusion/pull/10448

## Which issue does this PR close?

Closes https://github.com/apache/datafusion/issues/10291

## Rationale for this change

Follow on to https://github.com/apache/datafusion/pull/10366 from @dmitrybugakov  ❤️ 

Make planning faster by removing clones in filter pushdown

## What changes are included in this PR?
Rewrite the `PushDownFilter` to not copy things 

## Are these changes tested?
Functionally it is covered by existing tests (specifically there are unit tests in the module, the sqllogictest and tpcds planner tests)

## Are there any user-facing changes?

There is no functional change

Performance improvement shows 4-6% faster for TPCH/TPDS planning

<details><summary>Details</summary>
<p>

```
++ critcmp main push_down_filter
group                                         main                                   push_down_filter
-----                                         ----                                   ----------------
logical_aggregate_with_join                   1.00  1212.9±18.64µs        ? ?/sec    1.00  1217.1±10.78µs        ? ?/sec
logical_plan_tpcds_all                        1.00    159.3±1.73ms        ? ?/sec    1.00    158.7±1.79ms        ? ?/sec
logical_plan_tpch_all                         1.01     17.0±0.19ms        ? ?/sec    1.00     16.9±0.18ms        ? ?/sec
logical_select_all_from_1000                  1.01     18.8±0.14ms        ? ?/sec    1.00     18.7±0.09ms        ? ?/sec
logical_select_one_from_700                   1.00    811.6±7.82µs        ? ?/sec    1.01   820.1±29.24µs        ? ?/sec
logical_trivial_join_high_numbered_columns    1.00    760.1±8.61µs        ? ?/sec    1.00   762.9±13.46µs        ? ?/sec
logical_trivial_join_low_numbered_columns     1.00    744.4±7.41µs        ? ?/sec    1.01    749.5±8.29µs        ? ?/sec
physical_plan_tpcds_all                       1.06   1359.1±7.59ms        ? ?/sec    1.00   1284.5±8.00ms        ? ?/sec
physical_plan_tpch_all                        1.04     92.1±1.28ms        ? ?/sec    1.00     88.3±1.23ms        ? ?/sec
physical_plan_tpch_q1                         1.09      5.1±0.08ms        ? ?/sec    1.00      4.7±0.08ms        ? ?/sec
physical_plan_tpch_q10                        1.05      4.4±0.09ms        ? ?/sec    1.00      4.2±0.08ms        ? ?/sec
physical_plan_tpch_q11                        1.07      4.0±0.08ms        ? ?/sec    1.00      3.7±0.07ms        ? ?/sec
physical_plan_tpch_q12                        1.05      3.1±0.05ms        ? ?/sec    1.00      2.9±0.05ms        ? ?/sec
physical_plan_tpch_q13                        1.04      2.1±0.04ms        ? ?/sec    1.00      2.1±0.04ms        ? ?/sec
physical_plan_tpch_q14                        1.10      2.8±0.05ms        ? ?/sec    1.00      2.5±0.05ms        ? ?/sec
physical_plan_tpch_q16                        1.05      3.8±0.09ms        ? ?/sec    1.00      3.6±0.07ms        ? ?/sec
physical_plan_tpch_q17                        1.01      3.6±0.06ms        ? ?/sec    1.00      3.5±0.08ms        ? ?/sec
physical_plan_tpch_q18                        1.04      4.0±0.06ms        ? ?/sec    1.00      3.8±0.09ms        ? ?/sec
physical_plan_tpch_q19                        1.11      6.3±0.07ms        ? ?/sec    1.00      5.7±0.08ms        ? ?/sec
physical_plan_tpch_q2                         1.05      7.9±0.06ms        ? ?/sec    1.00      7.6±0.07ms        ? ?/sec
physical_plan_tpch_q20                        1.01      4.5±0.09ms        ? ?/sec    1.00      4.5±0.09ms        ? ?/sec
physical_plan_tpch_q21                        1.05      6.2±0.07ms        ? ?/sec    1.00      6.0±0.09ms        ? ?/sec
physical_plan_tpch_q22                        1.04      3.4±0.07ms        ? ?/sec    1.00      3.3±0.07ms        ? ?/sec
physical_plan_tpch_q3                         1.05      3.2±0.06ms        ? ?/sec    1.00      3.0±0.05ms        ? ?/sec
physical_plan_tpch_q4                         1.05      2.4±0.05ms        ? ?/sec    1.00      2.2±0.04ms        ? ?/sec
physical_plan_tpch_q5                         1.04      4.5±0.06ms        ? ?/sec    1.00      4.4±0.07ms        ? ?/sec
physical_plan_tpch_q6                         1.07  1615.0±89.07µs        ? ?/sec    1.00  1505.8±20.04µs        ? ?/sec
physical_plan_tpch_q7                         1.03      5.8±0.07ms        ? ?/sec    1.00      5.6±0.07ms        ? ?/sec
physical_plan_tpch_q8                         1.05      7.4±0.08ms        ? ?/sec    1.00      7.1±0.09ms        ? ?/sec
physical_plan_tpch_q9                         1.06      5.7±0.08ms        ? ?/sec    1.00      5.4±0.07ms        ? ?/sec
physical_select_all_from_1000                 1.01     61.8±1.40ms        ? ?/sec    1.00     61.1±0.27ms        ? ?/sec
physical_select_one_from_700                  1.00      3.7±0.04ms        ? ?/sec    1.00      3.7±0.05ms        ? ?/sec```

</p>
</details>

### CODE REVIEW

**Comment by alamb** [comment]
@jackwener I know this is a fairly large PR, but I think you were the one who did the last major rework of this code. Do you happen to have any time to review this PR?

**Review by jackwener — APPROVED** [review]
A very nice job to me! Thank you @alamb .
Besides enhancing performance, I am pleased to see that the readability of this part of the code has significantly improved. During the PR review process, I found the entire workflow to be much clearer.

File: datafusion/optimizer/src/push_down_filter.rs [hunk 1]
```diff
@@ -941,22 +935,65 @@ impl OptimizerRule for PushDownFilter {
...
+///
+/// Ending:
+/// ```text
+/// plan
+///   new_child
+/// ```
+fn insert_below(
+    plan: LogicalPlan,
+    new_child: LogicalPlan,
+) -> Result<Transformed<LogicalPlan>> {
```
[the remark this answers is not in the document]
Comment (alamb) [reply datafusion/optimizer/src/push_down_filter.rs:980]: Maybe I should pull it into optimizer/utils 🤔

**Review by alamb — COMMENTED** [review]
Thank you very much for the review @jackwener  🙏

**Pull request merged**